File: Search_UnknownLength.java
// Time Complexity : O(logn) 
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

// Approach:

// Got the low and high values by putting a while loop on a condition that element at high < target
// Applied binary search to get the answer
/**
 * // This is ArrayReader's API interface.
 * // You should not implement it, or speculate about its implementation
 * interface ArrayReader {
 * public int get(int index) {}
 * }
 */

class Solution {
    public int search(ArrayReader reader, int target) {
        int low = 0, high = 1;
        while (reader.get(high) < target) {
            low = high;
            high = 2 * high;
        }
        // System.out.println(low);
        // System.out.println(high);
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (reader.get(mid) == target)
                return mid;
            else if (reader.get(mid) < target)
                low = mid + 1;
            else
                high = mid - 1;
        }
        return -1;
    }
}

File: searchUnknownLength.py
# Time Complexity : O(logN)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# """
# This is ArrayReader's API interface.
# You should not implement it, or speculate about its implementation
# """
#class ArrayReader:
#    def get(self, index: int) -> int:

class Solution:
    def search(self, reader, target):
        """
        :type reader: ArrayReader
        :type target: int
        :rtype: int
        """
        low = 0
        high = 1
        # Find the upper limit where we can use as the high pointer.
        while reader.get(high) < target:
            high = high * 2

        # Apply binary search as we have got the low and high values now.
        while low <= high:
            mid = (low+high)//2
            val = reader.get(mid)
            if val == target:
                return mid
            if val > target:
                high = mid - 1
            else:
                low = mid + 1
        return -1